src/ctap/pin_protocol_v1.rs

```diff
@@ -0,0 +1,1137 @@
+    let mut pin = vec![];
+    'pin_block_loop: for block in blocks.iter().take(PIN_PADDED_LENGTH / 16) {
+        for cur_char in block.iter() {
+            if *cur_char != 0 {
+                pin.push(*cur_char);
+            } else {
+                break 'pin_block_loop;
+            }
+        }
+    }
+    #[cfg(feature = "with_ctap2_1")]
+    let min_pin_length = persistent_store.min_pin_length() as usize;
+    #[cfg(not(feature = "with_ctap2_1"))]
+    let min_pin_length = 4;
+    if pin.len() < min_pin_length || pin.len() == PIN_PADDED_LENGTH {
+        // TODO(kaczmarczyck) check 4 code point minimum instead
+        // TODO(kaczmarczyck) check last byte == 0x00
+        return false;
+    }
+    let mut pin_hash = [0u8; 16];
+    pin_hash.copy_from_slice(&Sha256::hash(&pin[..])[..16]);
+    persistent_store.set_pin_hash(&pin_hash);
+    true
+}
+
+#[cfg(feature = "with_ctap2_1")]
+#[cfg_attr(test, derive(IntoEnumIterator))]
+// TODO remove when all variants are used
+#[allow(dead_code)]
+pub enum PinPermission {
+    // All variants should use integers with a single bit set.
+    MakeCredential = 0x01,
+    GetAssertion = 0x02,
+    CredentialManagement = 0x04,
+    BioEnrollment = 0x08,
+    PlatformConfiguration = 0x10,
+    AuthenticatorConfiguration = 0x20,
+}
+
+pub struct PinProtocolV1 {
+    key_agreement_key: crypto::ecdh::SecKey,
+    pin_uv_auth_token: [u8; PIN_TOKEN_LENGTH],
+    consecutive_pin_mismatches: u8,
+    #[cfg(feature = "with_ctap2_1")]
+    permissions: u8,
+    #[cfg(feature = "with_ctap2_1")]
+    permissions_rp_id: Option<String>,
+}
+
+impl PinProtocolV1 {
+    pub fn new(rng: &mut impl Rng256) -> PinProtocolV1 {
+        let key_agreement_key = crypto::ecdh::SecKey::gensk(rng);
+        let pin_uv_auth_token = rng.gen_uniform_u8x32();
+        PinProtocolV1 {
+            key_agreement_key,
+            pin_uv_auth_token,
+            consecutive_pin_mismatches: 0,
+            #[cfg(feature = "with_ctap2_1")]
+            permissions: 0,
+            #[cfg(feature = "with_ctap2_1")]
+            permissions_rp_id: None,
+        }
+    }
+
+    fn check_pin_hash_enc(
+        &mut self,
+        rng: &mut impl Rng256,
+        persistent_store: &mut PersistentStore,
+        aes_dec_key: &crypto::aes256::DecryptionKey,
+        pin_hash_enc: Vec<u8>,
+    ) -> Result<(), Ctap2StatusCode> {
+        match persistent_store.pin_hash() {
+            Some(pin_hash) => {
+                if self.consecutive_pin_mismatches >= 3 {
+                    return Err(Ctap2StatusCode::CTAP2_ERR_PIN_AUTH_BLOCKED);
+                }
+                // We need to copy the pin hash, because decrementing the pin retries below may
+                // invalidate the reference (if the page containing the pin hash is compacted).
+                let pin_hash = pin_hash.to_vec();
+                persistent_store.decr_pin_retries();
+                if pin_hash_enc.len() != PIN_AUTH_LENGTH {
```

Comment:
Should we check the length of `pin_hash` as well?

Reply:
It is loaded from persistent storage (where it should only ever get with 16 bytes), and later input into `array_ref` and `ct_eq`. So we would change behaviour from panic to outputting an OpenSK-specific error code. Is this change in behaviour what you were asking for?

Reply:
Ah, now I think I see the problem. The `to_vec()` should be a `clone()`. The `persistent_store` already returns an `Option<&[u8; PIN_AUTH_LENGTH]>`. I also removed the `array_ref` here. I think returning results in the storage  is planned and I'll not touch that here.

Reply:
Why would you need to clone it? Isn't a reference enough? I only see it used in `array_ref` again.

Reply:
The clone is necessary as explained in the comment above the line. We need the following order of events:

1. get `pin_hash` from persistent store
2. decrease PIN retries in persistent store
3. use `pin_hash`

The 2. step could invalidate the reference returned in step 1. I could change the match on `pin_hash` into an early returning assignment, to make the flow clearer?

Reply:
Hum, it wasn't clear to me that it was needed due to a borrow checker violation. The comment could make it clearer, e.g. `because decrementing the pin retries below mutably borrows from the same persistent_store reference, and therefore invalidates the pin_hash reference`.

It's not exactly that it "*may* invalidate the reference", but rather that the reference is automatically and *always* invalidated by Rust's borrow checker.

In any case, we shouldn't call `.to_vec()` to clone a `[u8; PIN_AUTH_LENGTH]`. With a `.clone()` it should be fine.

Reply:
I still don't see why we need to check the length. Don't we get back a statically sized slice/array, as written in https://github.com/google/OpenSK/pull/127/files#r463190693?

Reply:
Never mind, I misread it as the other "pin hash".